### PR TITLE
Support nested member-access in filter expressions (#392) [v4]

### PR DESCRIPTION
**Pull request by ltwlf**
## Summary
- Function-based identity keys returning nested paths (e.g. `positionNumber.value`) now produce `@.positionNumber.value` (RFC 9535) instead of `@['positionNumber.value']`
- String-based identity keys with dots still use bracket notation `@['c.d']`
- Adds `embeddedKeyIsPath` flag to `IChange`
- Bumps to `4.10.3`

## Test plan
- [x] 106 v4 tests pass, lint/typecheck clean

### CODE REVIEW

File: src/jsonDiff.ts [hunk 1]
```diff
@@ -653,13 +663,21 @@ const removeKey = (obj: any, key: any, embeddedKey: any) => {
   }
 };
 
-const indexOfItemInArray = (arr: any[], key: any, value: any) => {
+/** Resolve a property on an object. When isPath is true, traverses nested dot-separated segments. */
+const resolveProperty = (obj: any, key: string, isPath?: boolean): any => {
+  if (!isPath || !key.includes('.')) return obj[key];
```
[the remark this answers is not in the document]
Comment (ltwlf) [reply src/jsonDiff.ts]: Fixed in c73e893 — `resolveProperty` returns `undefined` for null/undefined objects.

File: src/jsonDiff.ts [hunk 2]
```diff
@@ -819,7 +840,10 @@ const revertArrayChange = (arr: any[], change: any) => {
           element = arr[index];
         }
       } else {
-        element = arr.find((el) => el[change.embeddedKey]?.toString() === subchange.key.toString());
+        element = arr.find((el) => {
+          const resolved = resolveProperty(el, change.embeddedKey, change.embeddedKeyIsPath);
+          return resolved != null && String(resolved) === String(subchange.key);
+        });
```
[the remark this answers is not in the document]
Comment (ltwlf) [reply src/jsonDiff.ts]: Fixed in c73e893 — `isPath` threaded through `revertLeafChange` → `removeKey`.

File: src/jsonDiff.ts [hunk 3]
```diff
@@ -564,12 +569,15 @@ const compareArray = (oldObj: any, newObj: any, path: any, keyPath: any, options
...
   if (diffs.length) {
+    const resolvedKey = isFunctionKey ? uniqKey(newObj[0] ?? oldObj[0], true) : uniqKey;
     return [
       {
         type: Operation.UPDATE,
         key: getKey(path),
-        embeddedKey: typeof uniqKey === 'function' && uniqKey.length === 2 ? uniqKey(newObj[0], true) : uniqKey,
+        embeddedKey: resolvedKey,
+        ...(isFunctionKey && typeof resolvedKey === 'string' && resolvedKey.includes('.') ? { embeddedKeyIsPath: true } : {}),
         changes: diffs
```
[the remark this answers is not in the document]
Comment (ltwlf) [reply src/jsonDiff.ts]: Fixed in eff3ec5 — same `NESTED_PATH_RE` gating in both `compareArray` and `unatomizeChangeset`. Edge case test added.

File: tests/atomizeChangeset.test.ts [hunk 4]
```diff
@@ -62,38 +62,73 @@ describe('atomizeChangeset', () => {
...
   });
 
-  test('when atomizing and unatomizing with bracket-notation filter keys', () => {
+  test('when function-based key returns non-identifier dotted name, uses bracket notation', () => {
+    const oldObj = { items: [{ 'a-b': { c: 'x' }, v: 1 }] };
+    const newObj = { items: [{ 'a-b': { c: 'x' }, v: 2 }] };
+
+    const resolver: FunctionKey = (obj, shouldReturnKeyName) => {
+      if (shouldReturnKeyName) return "a-b.c";
+      return obj['a-b']?.c;
```
[the remark this answers is not in the document]
Comment (ltwlf) [reply tests/atomizeChangeset.test.ts]: This test verifies the path format — `a-b.c` fails `NESTED_PATH_RE` so it gets bracket notation. Apply for this combination is a known limitation (the key name doesn't match the data structure). The test documents the expected output, not round-trip.
